### ISSUE TITLE
Commit 1c2ae10dc057d5c7c76d0df2c5a92fc753e4769a causes breakage

### ISSUE THREAD

**Issue by nbougalis** [XRPLF/rippled#4601]
Commit 1c2ae10dc057d5c7c76d0df2c5a92fc753e4769a causes breakage: the way it checks for certain flags breaks, resulting in errors. It should be reverted.

**Comment by intelliot** [comment]
Can you share a demonstration of the breakage (and errors)?

**Comment by nbougalis** [comment]
Consider the function `conditionMet`, specifically lines 121 and 122 and 146-147 of Handler.h: 

https://github.com/XRPLF/rippled/commit/1c2ae10dc057d5c7c76d0df2c5a92fc753e4769a#diff-41d8d23b15cb894b075204dcd88d7bf9ca223fa77f43e28bb22cf38b805cabfcL121-L122

https://github.com/XRPLF/rippled/commit/1c2ae10dc057d5c7c76d0df2c5a92fc753e4769a#diff-41d8d23b15cb894b075204dcd88d7bf9ca223fa77f43e28bb22cf38b805cabfcL121-L122

The change is broken: where it previously, it checked for a _specific_ condition (e.g. `NEEDS_CLOSED_LEDGER`) and only entered the `if` block if it was met, it now executes that code if _any_ condition is set.

A prime example of a RPC request that fails is the `tx` command, which has `NEEDS_NETWORK_CONNECTION` and will now not work unless the server has _both_ a current and a closed ledger.

The problem isn't the powers of two change. It's how the checks where restructured.

**Comment by intelliot** [comment]
Is it possible to write a unit test that detects the error?

**Comment by ckeshava** [comment]
hello Nik, thank you very much for pointing this out. I'm working on this issue.

**Comment by ckeshava** [comment]
I will revert the changes to the condition evaluation.

The existing unit tests do not catch this error. It appears that there are no unit tests for `https://github.com/XRPLF/rippled/blob/develop/src/ripple/rpc/impl/Handler.h`, but this code is indirectly being tested in `https://github.com/XRPLF/rippled/blob/develop/src/test/rpc/RPCCall_test.cpp`. 

Personally, I feel it is better to add a new unit test file for the `Handler.h` (because it would have a smaller granularity) instead of piling on additional tests in `RPCCall_test`. Do you guys think that is okay?

**Comment by ximinez** [comment]
> A prime example of a RPC request that fails is the `tx` command, which has `NEEDS_NETWORK_CONNECTION` and will now not work unless the server has _both_ a current and a closed ledger.

I don't think there's a behavior change at all. I commented on the original PR to that effect: https://github.com/XRPLF/rippled/pull/4239#discussion_r1224782464

Here's more detail.

In the old code,

```
NO_CONDITION == 0
NEEDS_NETWORK_CONNECTION == 1 == 0b0001
NEEDS_CURRENT_LEDGER     == 3 == 0b0011
NEEDS_CLOSED_LEDGER      == 5 == 0b0101
```

Consider the condition on line 122:

```
condition_required & NEEDS_NETWORK_CONNECTOIN
```
which is equivalent to 
```
condition_required & 0b0001
```

`condition_required` can be any of those 4 values.

| `condition_required` | `condition_required & NEEDS_NETWORK_CONNECTION` | boolean result |
|-|-|-|
| `NO_CONDITION == 0` | `0` | `false` |
| `NEEDS_NETWORK_CONNECTION == 0b0001` | `0b0001` | `true` |
| `NEEDS_CURRENT_LEDGER == 0b0011` | `0b0001` | `true` |
| `NEEDS_CLOSED_LEDGER == 0b0101` | `0b0001` | `true` |

Basically, for any value except `NO_CONDITION` (0), the expression evaluates to `true`.

The new values are 0, 1, 2, and 4, respectively, but the result is the same. `condition_required != NO_CONDITION` in the new code is functionally equivalent to `condition_required & NEEDS_NETWORK_CONNECTION` in the old code.

It should be clear that the results are the same for the other values.

**Comment by ckeshava** [comment]
Thanks for the analysis Ed.

For the specific objections raised by Nik (on lines 121, 146), here is a similar truth table for comparison with `NEEDS_CURRENT_LEDGER` and `NEEDS_CLOSED_LEDGER`.  Both of these truth tables are identical to `condition_required != NO_CONDITION`. 

`condition_required` | `condition_required & NEEDS_CURRENT_LEDGER` | boolean result
-- | -- | --
`NO_CONDITION` == 0 | 0 | `false`
`NEEDS_NETWORK_CONNECTION` == 0b0001 | 0b0001 | `true`
`NEEDS_CURRENT_LEDGER` == 0b0011 | 0b0011 | `true`
`NEEDS_CLOSED_LEDGER` == 0b0101 | 0b0001 | `true`



`condition_required` | `condition_required & NEEDS_CLOSED_LEDGER` | boolean result
-- | -- | --
`NO_CONDITION` == 0 | 0 | `false`
`NEEDS_NETWORK_CONNECTION` == 0b0001 | 0b0001 | `true`
`NEEDS_CURRENT_LEDGER` == 0b0011 | 0b0001 | `true`
`NEEDS_CLOSED_LEDGER` == 0b0101 | 0b0101 | `true`

**Comment by nbougalis** [comment]
So, you're right. In looking at the original commit, I missed the part where `NEEDS_CURRENT_LEDGER` and `NEEDS_CLOSED_LEDGER` are compound flags (i.e. they're set to `NEEDS_NETWORK_CONNECTION | something`).

While reading your posts above and looking more closely, I realized that this PR didn't break things. Instead, it simply replicated the _existing_ breakage.

Let's examine a particular instance of this breakage. The old code would do:

```
 if (context.app.getOPs().isAmendmentBlocked() &&
        (condition_required & NEEDS_CURRENT_LEDGER ||
         condition_required & NEEDS_CLOSED_LEDGER))
```

There are commands (e.g. `ledger_cleaner` or `tx`) which just need a network connection and nothing more (i.e. they don't require a current or closed ledger!). Let's consider what happens when processing such a command.

The check reduces to: `(NEEDS_NETWORK_CONNECTION & NEEDS_CURRENT_LEDGER) || (NEEDS_NETWORK_CONNECTION & NEEDS_CLOSED_LEDGER)`, which, in turn reduces to: `(1 & 3) || (1 & 5)`, which reduces to `1 || 1` aka `true.

Because of the way `NEEDS_CURRENT_LEDGER` and `NEEDS_CLOSED_LEDGER` are defined and the semantics of the bitwise `AND` operator, we'll enter that `if` even for commands which **don't** require a current or a closed ledger. Oops.

In the new code, we simply check `condition_required != NO_CONDITION`. I agree that, given the definition of the permissions in the old code, the new code is equivalent. But it's still wrong.

Also note, that the code is equivalent _by chance_ and the new design is flaky. If a new permission level was introduced (one that, perhaps, doesn't "imply" `NEEDS_NETWORK_CONNECTION`) the existing checks against `NO_CONDITION` might be very wrong, and it might not be immediately obvious that that's the case.

Anyways.

This isn't a disaster; it's just that some RPC operations can't be executed when they should.

**Comment by ximinez** [comment]
@nbougalis I agree with your analysis and conclusions. [I made a similar comment on the original PR](https://github.com/XRPLF/rippled/pull/4239#discussion_r1224782464), but left it as more of an open question, instead of pushing it to an answer.

So I don't think the solution is to rollback the original fix entirely. However, it might be worth changing the relevant checks back to check the original conditions, then audit the affected functions to see if the conditions makes sense. IMHO, if the conditions are correct, this can only be an improvement, since it will allow some RPC operations to execute where they previously wouldn't.
